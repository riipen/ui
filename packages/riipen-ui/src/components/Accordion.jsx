import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import AccordionContext from "../contexts/AccordionContext";
import ThemeContext from "../styles/ThemeContext";

import Collapse from "./Collapse";

const Accordion = props => {
  const {
    children: childrenProp,
    classes,
    defaultExpanded,
    disabled,
    ...other
  } = props;

  const theme = React.useContext(ThemeContext);

  const [expanded, setExpanded] = React.useState(defaultExpanded);

  React.useEffect(() => {
    setExpanded(defaultExpanded);
  }, [defaultExpanded]);

  const handleChange = React.useCallback(() => {
    setExpanded(!expanded);
  }, [expanded, setExpanded]);

  const [summary, ...children] = React.Children.toArray(childrenProp);

  const contextValue = React.useMemo(
    () => ({ disabled, expanded, toggle: handleChange }),
    [expanded, disabled, handleChange]
  );

  return (
    <div
      className={clsx(
        "root",
        {
          expanded,
          disabled
        },
        classes
      )}
      {...other}
    >
      <AccordionContext.Provider value={contextValue}>
        {summary}
      </AccordionContext.Provider>
      <Collapse in={expanded}>
        <div
          aria-labelledby={summary.props.id}
          id={summary.props["aria-controls"]}
        >
          {children}
        </div>
      </Collapse>

      <style jsx>{`
        .root {
          position: relative;
          transition: ${theme.transitions.create(["margin"])};
          overflow-anchor: none;
        }
        .root:before {
          position: absolute;
          left: 0;
          top: -1;
          right: 0;
          height: 1;
          content: "";
          opacity: 1;
          background-color: ${theme.palette.divider};
          transition: ${theme.transitions.create([
            "background-color",
            "opacity"
          ])};
        }
        .root:first-child:before {
          display: "none";
        }

        .root.expanded:before {
          opacity: 0;
        }
      `}</style>
    </div>
  );
};

Accordion.propTypes = {
  /**
   * The content of the accordion.
   */
  children: PropTypes.node.isRequired,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * If `true`, expands the accordion by default.
   */
  defaultExpanded: PropTypes.bool,

  /**
   * If `true`, the accordion will be displayed in a disabled state.
   */
  disabled: PropTypes.bool
};

Accordion.defaultProps = {
  classes: [],
  defaultExpanded: false,
  disabled: false
};

Accordion.displayName = "Accordion";

export default Accordion;
