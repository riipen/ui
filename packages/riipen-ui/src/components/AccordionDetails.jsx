import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

const AccordionDetails = props => {
  const { classes, ...other } = props;

  return (
    <React.Fragment>
      <div className={clsx("root", classes)} {...other} />
      <style jsx>{`
        .root {
          display: flex;
        }
      `}</style>
    </React.Fragment>
  );
};

AccordionDetails.propTypes = {
  /**
   * The content of the accordion details.
   */
  children: PropTypes.node,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array
};

AccordionDetails.displayName = "AccordionDetails";

export default AccordionDetails;
