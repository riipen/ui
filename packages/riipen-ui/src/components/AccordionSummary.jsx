import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import css from "styled-jsx/css";

import ThemeContext from "../styles/ThemeContext";

import AccordionContext from "./AccordionContext";
import ButtonIcon from "./ButtonIcon";

const AccordionSummary = props => {
  const { children, classes, iconProps, icon, ...other } = props;

  const theme = React.useContext(ThemeContext);

  const [focusedState, setFocusedState] = React.useState(false);

  const getLinkedStyles = () => {
    return css.resolve`
      .root {
        align-items: center;
        cursor: pointer;
        display: flex;
        transition: ${theme.transitions.create([
          "background-color",
          "min-height"
        ])};
        user-select: none;
      }
      .root.disabled {
        cursor: initial;
      }

      .content {
        flex-grow: 1;
        transition: ${theme.transitions.create(["margin"])};
      }

      .icon {
        transform: rotate(0deg);
        transition: ${theme.transitions.create(["transform"])};
      }
      .icon.expanded {
        transform: rotate(180deg);
      }
    `;
  };

  const handleFocusVisible = () => {
    setFocusedState(true);
  };

  const handleBlur = () => {
    setFocusedState(false);
  };

  const { disabled = false, expanded, toggle } = React.useContext(
    AccordionContext
  );

  const handleClick = event => {
    if (toggle && !disabled) {
      toggle(event);
    }
  };

  const linkedStyles = getLinkedStyles();

  return (
    <div
      disabled={disabled}
      aria-expanded={expanded}
      className={clsx(
        linkedStyles.className,
        "root",
        {
          disabled,
          expanded,
          focused: focusedState
        },
        classes
      )}
      onClick={handleClick}
      onFocusVisible={handleFocusVisible}
      onBlur={handleBlur}
      {...other}
    >
      <div
        className={clsx(linkedStyles.className, "content", {
          expanded
        })}
      >
        {children}
      </div>
      {icon && (
        <ButtonIcon
          component="div"
          {...iconProps}
          classes={[
            clsx(
              linkedStyles.className,
              "icon",
              {
                expanded
              },
              iconProps.classes
            )
          ]}
          disabled={disabled}
        >
          {icon}
        </ButtonIcon>
      )}
      {linkedStyles.styles}
    </div>
  );
};

AccordionSummary.propTypes = {
  /**
   * The content of the accordion summary.
   */
  children: PropTypes.node,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * The icon to display as the expand indicator.
   */
  icon: PropTypes.node,

  /**
   * Props applied to the `ButtonIcon` element wrapping the expand icon.
   */
  iconProps: PropTypes.object
};

AccordionSummary.defaultProps = {
  classes: [],
  iconProps: {}
};

AccordionSummary.displayName = "AccordionSummary";

export default AccordionSummary;
