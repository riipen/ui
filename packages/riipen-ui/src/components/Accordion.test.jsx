import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Accordion from "./Accordion";
import AccordionDetails from "./AccordionDetails";
import AccordionSummary from "./AccordionSummary";

describe("<Accordion>", () => {
  it("renders without errors", () => {
    let error;
    const child = <AccordionDetails />;

    try {
      mount(<Accordion>{child}</Accordion>);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const summary = <AccordionSummary />;
    const child = <AccordionDetails />;

    const wrapper = mount(
      <Accordion>
        {summary}
        {child}
      </Accordion>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const child = <AccordionDetails />;

      const wrapper = mount(<Accordion>{child}</Accordion>);

      expect(wrapper.find("Accordion").props().defaultExpanded).toEqual(false);
      expect(wrapper.find("Accordion").props().disabled).toEqual(false);
    });
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <AccordionDetails>hello</AccordionDetails>;

      const wrapper = mount(<Accordion>{child}</Accordion>);

      expect(wrapper.find("AccordionDetails").props().children).toEqual(
        "hello"
      );
    });
  });

  describe("classes prop", () => {
    it("applies custom classes", () => {
      const classes = ["classOne"];
      const child = <AccordionDetails>hello</AccordionDetails>;

      const wrapper = mount(<Accordion classes={classes}>{child}</Accordion>);

      expect(
        wrapper
          .find("Accordion")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });
  });

  describe("defaultExpanded prop", () => {
    it("sets expanded class with given defaultExpanded", () => {
      const defaultExpanded = true;
      const child = <AccordionDetails>hello</AccordionDetails>;

      const wrapper = mount(
        <Accordion defaultExpanded={defaultExpanded}>{child}</Accordion>
      );

      expect(wrapper.find("Collapse").props().in).toEqual(defaultExpanded);
      expect(
        wrapper
          .find("Accordion")
          .childAt(0)
          .hasClass("expanded")
      ).toEqual(true);
    });
  });

  describe("expanded state", () => {
    it("sets correct expanded class when summary is clicked", () => {
      const defaultExpanded = true;
      const summary = <AccordionSummary>summary</AccordionSummary>;
      const child = <AccordionDetails>hello</AccordionDetails>;

      const wrapper = mount(
        <Accordion defaultExpanded={defaultExpanded}>
          {summary}
          {child}
        </Accordion>
      );

      wrapper.find("AccordionSummary").simulate("click");

      expect(wrapper.find("Collapse").props().in).toEqual(!defaultExpanded);
      expect(
        wrapper
          .find("Accordion")
          .childAt(0)
          .hasClass("expanded")
      ).toEqual(false);
    });
  });

  describe("disabled prop", () => {
    it("applies disabled class when true", () => {
      const disabled = true;
      const child = <AccordionDetails>hello</AccordionDetails>;

      const wrapper = mount(<Accordion disabled={disabled}>{child}</Accordion>);

      expect(
        wrapper
          .find("Accordion")
          .childAt(0)
          .hasClass("disabled")
      ).toEqual(true);
    });

    it("does not apply disabled class when false", () => {
      const disabled = false;
      const child = <AccordionDetails>hello</AccordionDetails>;

      const wrapper = mount(<Accordion disabled={disabled}>{child}</Accordion>);

      expect(
        wrapper
          .find("Accordion")
          .childAt(0)
          .hasClass("disabled")
      ).toEqual(false);
    });
  });
});
