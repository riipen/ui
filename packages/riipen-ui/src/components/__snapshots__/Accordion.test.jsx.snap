// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion> renders correct snapshot 1`] = `
<Accordion
  classes={Array []}
  defaultExpanded={false}
  disabled={false}
  provideContextToChildren={false}
>
  <div
    className="jsx-839250457 root"
  >
    <AccordionSummary
      classes={Array []}
      iconProps={Object {}}
      key=".0"
    >
      <div
        aria-expanded={false}
        className="jsx-3687637527 root"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="jsx-3687637527 content"
        />
        <JSXStyle
          dynamic={
            Array [
              "background-color 300ms cubic-bezier(0.4, 0, 0.2, 1), min-height 300ms cubic-bezier(0.4, 0, 0.2, 1)",
              "margin 300ms cubic-bezier(0.4, 0, 0.2, 1)",
              "transform 300ms cubic-bezier(0.4, 0, 0.2, 1)",
            ]
          }
          id="1644211938"
        />
      </div>
    </AccordionSummary>
    <Collapse
      classes={Array []}
      component="div"
      in={false}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={300}
        unmountOnExit={false}
      >
        <div
          className="jsx-2207988454 root exited"
        >
          <div>
            <div
              className="jsx-839250457"
            >
              <AccordionDetails
                key=".1"
              >
                <div
                  className="jsx-1644707333 root"
                />
                <JSXStyle
                  id="1644707333"
                />
              </AccordionDetails>
            </div>
          </div>
          <JSXStyle
            dynamic={
              Array [
                "height 300ms cubic-bezier(0.4, 0, 0.2, 1)",
              ]
            }
            id="4123931595"
          />
        </div>
      </Transition>
    </Collapse>
    <JSXStyle
      dynamic={
        Array [
          "margin 300ms cubic-bezier(0.4, 0, 0.2, 1)",
          "#ddd",
          "background-color 300ms cubic-bezier(0.4, 0, 0.2, 1), opacity 300ms cubic-bezier(0.4, 0, 0.2, 1)",
        ]
      }
      id="209282601"
    />
  </div>
</Accordion>
`;
